### PR TITLE
Water@Hanh Solo - Tasklist

### PR DESCRIPTION
**Pull request by seattlefurby17**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Model creates an entry to database with all the necessary information. 
Describe in your own words what the Controller is doing in Rails | Controller takes requests from users and response with data from Model
Describe in your own words what the View is doing in Rails | View displays information processed by Controller
Describe an edge-case controller test you wrote |I am not sure what would be an edge-case in this project.
What is the purpose of using strong params? (i.e. the params method in the controller) | In Rails, controllers are not permitted to pass parameters directly from the form into our ActiveRecord model instances. This is a security feature to ensure users cannot update sensitive model attributes that they should not be updating.
How are Rails migrations related to Rails models? | Rails migrations are the blue print of Rails models.
Describe one area of Rails that are still unclear on | I am not sure where and what to update based on the requirements, show page, index page, etc.

### CODE REVIEW

File: app/views/tasks/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  <p>Please provide the following information to add your new task:</p>
+
+  <%= f.label :name %>
+  <%= f.text_field :name %>
+
+  <%= f.label :description %>
+  <%= f.text_field :description %>
+
+   <%= f.label :completed_at %>
+  <%= f.text_field :completed_at %>
```
Comment (beccaelenzil) [inline app/views/tasks/new.html.erb:14]: Consider whether you want to input this in a form or default to something like `nil` when a task is created.

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    else
+      @task.update(
+        completed_at: nil
+      )
+      redirect_to tasks_path 
+      return
+    end
+  end
+
+  def mark_complete
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:61]: Currently as soon as you make a task `complete` is truthy so that the view says "unmark complete" even if there's nothing filled in for completed_at. This behavior seems not quite right.

Note that `completed_at` and `complete` are redundant attributes. Consider what should your code do if `completed` is `true`, but the `completed at` is `nil`? Instead you might say that a `nil` value for `completed_at` indicates the task is incomplete, and a non-`nil` value indicates it is complete.

**Review by beccaelenzil — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️ Good use of resources                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | This was not required, but consider for your next project!                                                            |
| Handles errors like nonexistant tasks              | ✔️                                                            |
| Uses `form_with` to render forms in Rails          | ✔️                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️    |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | ✔️ the mark complete behavior is a little off. See inline comments     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

Good work on task list! You've created your first web app in rails that follows rails conventions. Furthermore, you've hit the learning goals of working with partial views and strong params. I've left a couple inline comments for your to review. Keep up the hard work!


| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            |✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  | ✅    |
| Descriptive/Readable | ✅    |
| Concise              | ✅    |
